jupyter-org-extensions.el

```diff
@@ -0,0 +1,404 @@
+                              (let ((s (point)))
+                                (forward-line N)
+                                (buffer-substring s (point)))))
+                    (line-number-at-pos (match-beginning 0)))
+              p)))
+    (ivy-read "block: " (reverse p)
+              :action (lambda (candidate)
+                        (goto-char (point-min))
+                        (forward-line (1- (second candidate)))
+                        (outline-show-entry)
+                        (recenter)))))
+
+;;;###autoload
+(defun jupyter-org-jump-to-visible-block ()
+  "Jump to a visible src block with avy."
+  (interactive)
+  (avy-with jupyter-org-jump-to-block
+            (avy-jump "#\\+BEGIN_SRC" :beg (point-min) :end (point-max))))
+
+;;;###autoload
+(defun jupyter-org-edit-header ()
+  "Edit the src-block header in the minibuffer."
+  (interactive)
+  (let ((src-info (org-babel-get-src-block-info 'light)))
+    (unless src-info
+      (error "Not in a source block"))
+    (let* ((header-start (sixth src-info))
+           (header-end (save-excursion (goto-char header-start)
+                                       (line-end-position))))
+      (setf (buffer-substring header-start header-end)
+            (read-string "Header: "
+                         (buffer-substring header-start header-end))))))
+
+;;;###autoload
+(defun jupyter-org-kill-block-and-results ()
+  "Kill the block and its results."
+  (interactive)
+  (let ((src (org-element-context))
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-element-property :end src)))
+      (setq end (org-element-property :end src)))
+    (kill-region
+     (org-element-property :begin src)
+     end)))
+
+;;;###autoload
+(defun jupyter-org-copy-block-and-results ()
+  "Copy the src block at the current point and its results."
+  (interactive)
+  (let ((src (org-element-context))
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-babel-result-end)))
+      (setq end (org-element-property :end src)))
+    (kill-new
+     (buffer-substring
+      (org-element-property :begin src)
+      end))))
+
+;;;###autoload
+(defun jupyter-org-clone-block (&optional below)
+  "Clone the block above the current block.
+
+If BELOW is non-nil, add the cloned block below."
+  (interactive "P")
+  (let* ((src (org-element-context))
+         (code (org-element-property :value src)))
+    (unless (eq (org-element-type src) 'src-block)
+      (error "Not in a source block"))
+    (jupyter-org-insert-src-block below)
+    (delete-char 1)
+    (insert code)
+    ;; move to the end of the last line of the cloned block
+    (forward-line -1)
+    (end-of-line)))
+
+;;;###autoload
+(defun jupyter-org-merge-blocks ()
```

Comment:
This can be simplified by just working with the text in the buffer instead of pulling out all of the `:language`, `:switches`, etc. from the element. For example, we can delete and save the text in between the two source blocks using `delete-and-extract-region` on the region 

`((org-element-property :end current-src-block) . (org-element-property :begin next-src-block))`

then delete the `#+END_SRC ... #+BEGIN_SRC` and the two source blocks will be merged. Of course this is after deleting any results the source blocks might have. Then we move to the end of the merged source block and insert the saved text that was between the two.

Doing this also avoids having to consider if the source block delimiters are uppercase or lowercase as well.

WDYT?

Reply:
I think the logic behind this idea would be rather ugly to implement:
The idea is to be able to merge blocks next to each other as well as those that have org elements in between. For example, if there is text in between two blocks, it's still desirable to be able to merge them together without touching the org elements in the middle.

Reply:
Yes your right it doesn't make sense to delete the intermediate text since there could also be other stuff associated with the text, like overlays, that would be removed. I was thinking since the end result would be the same, it wouldn't matter that we messed with the intermediate text.

Below is a different take on the idea that just deletes the source block after the current block. I believe it is simpler than the `org-element` solution since it only deletes the contents of the source block after the current one instead of also deleting the current one.

```emacs-lisp
(defun jupyter-org-merge-blocks ()
  "Merge the current block with the next block."
  (interactive)
  (unless (org-in-src-block-p)
    (error "Not in a source block"))
  (let ((asrc (org-element-context)))
    (org-babel-remove-result)
    (org-babel-next-src-block)
    (let* ((bsrc (prog1 (org-element-context)
                   (org-babel-remove-result)))
           (bsrc-beg (set-marker
                      (make-marker)
                      (org-element-property :begin bsrc)))
           (bsrc-end (set-marker
                      (make-marker)
                      (jupyter-org-element-end-before-blanks bsrc))))
      (goto-char (jupyter-org-element-end-before-blanks asrc))
      (forward-line -1)
      (insert
       (delete-and-extract-region
        (save-excursion
          (goto-char (jupyter-org-element-begin-after-affiliated bsrc))
          (forward-line 1)
          (point))
        (save-excursion
          (goto-char bsrc-end)
          (forward-line -1)
          (point))))
      (delete-region bsrc-beg bsrc-end)
      (set-marker bsrc-beg nil)
      (set-marker bsrc-end nil))))
```

Reply:
Your proposal looks good.
I've modified to delete one extra line that was leftover and to move to the end of the last line.